nemo/core/classes/modelPT.py

```diff
@@ -46,9 +45,29 @@ class ModelPT(LightningModule, Model):
     Interface for Pytorch-lightning based NeMo models
     """
 
-    def __init__(self, cfg: ModelPTConfig = None, trainer=None):
+    def save_to(self, save_path: str):
+        """Saves model into .nemo format - includes both weights and configuration."""
+        pass
+
+    @classmethod
+    def restore_from(cls, restore_path: str):
+        """Fully (weights and structure) instantiates model from .nemo format"""
+        pass
+
+    def __init__(self, cfg: Union[ModelPTConfig, DictConfig, Dict] = None, trainer=None):
         super().__init__()
-        self._cfg = cfg
+        # First step is to make sure that self._cfg is DictConfig
+        if not (isinstance(cfg, ModelPTConfig) or isinstance(cfg, DictConfig) or isinstance(cfg, Dict)):
+            raise ValueError(
+                f"cfg argument must be of type ModelPTconfig, DictConfig or Dict. But got {type(cfg)} instead."
+            )
+        if isinstance(cfg, DictConfig):
+            self._cfg = cfg
+        else:
+            if dataclasses.is_dataclass(cfg):
+                cfg = dataclasses.asdict(cfg)
+
+            self._cfg = OmegaConf.create(cfg)
         self.save_hyperparameters(self._cfg)
```

Comment:
Important Note: We need to remove this `self.save_hyperparameters(self._cfg)` line. The LightningModel.hparams is used for logging more than anything - and neither Tensorflow nor other loggers support nested dictionaries for logging. On the other hand, pushing the immutable DictConfig inside this means that we cannot update / inject train time parameters - lr, loss value etc since these keys dont exist inside self._cfg anymore. Even if we down cast to dict each iteration, push these values and then convert back to DictConfig (horribly expensive operation per iteration), we cannot actually export this since during deserialization, it would find wrong keys like "lr" and "train_loss" etc.

So please remove this and directly use self._cfg in order to serialize / deserialize.

Reply:
hmmmm, @okuchaiev please correct me, but this was not for serialization/deserialization.

If I understood that added you added this as wanted to display params per a given experiment in tensorboard - and it worked.

Not sure how it will handle e.g. schedulers changing learning rate though...

Reply:
discussed